### PR TITLE
Implement Tuple operation directly in the compiler

### PR DESCRIPTION
**Pull request by nicolasstucki**
In this PR we gain performance of compiling code with tuple operations. The improvements are split into two.

First, introduce some optimizations on the generated code to avoid generating code that is too large.

Second, implement all optimizations and check in the new `TupleOptimizations` miniphase. It performs the equivalent optimizations as the first part. It expands the later in the compilation pipeline which saves time in phases before `TupleOptimizations` (placed just before `Erasure`).

### CODE REVIEW

**Comment by nicolasstucki** [comment]
test performance with #tuples please

**Comment by nicolasstucki** [comment]
test performance with #tuples please

**Comment by nicolasstucki** [comment]
test performance with #tuples please

**Comment by nicolasstucki** [comment]
test performance with #tuples please

**Comment by nicolasstucki** [comment]
test performance with #tuples please

**Comment by nicolasstucki** [comment]
test performance with #tuples please

**Comment by nicolasstucki** [comment]
test performance with #tuples please

**Comment by nicolasstucki** [comment]
test performance with #tuples please

**Comment by nicolasstucki** [comment]
test performance with #tuples please

**Comment by nicolasstucki** [comment]
test performance with #tuples please

**Comment by nicolasstucki** [comment]
test performance with #tuples please

File: compiler/src/dotty/tools/dotc/transform/TupleOptimizations.scala [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+import scala.annotation.tailrec
+
+/** Optimize generic operations on tuples */
+class TupleOptimizations extends MiniPhase with IdentityDenotTransformer {
+  import tpd._
+
+  def phaseName: String = "genericTuples"
+
+  override def transformApply(tree: tpd.Apply)(implicit ctx: Context): tpd.Tree = {
+    if (tree.symbol == defn.DynamicTuple_dynamicCons) transformTupleCons(tree)
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/transform/TupleOptimizations.scala]: To make this faster, put everything in
```
   if (tree.symbol.owner == defn.DynamicTupleModule) ...
```

File: compiler/src/dotty/tools/dotc/transform/TupleOptimizations.scala [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+    else if (tree.symbol == defn.DynamicTuple_dynamicSize) transformTupleSize(tree)
+    else if (tree.symbol == defn.DynamicTuple_dynamicConcat) transformTupleConcat(tree)
+    else if (tree.symbol == defn.DynamicTuple_dynamicApply) transformTupleApply(tree)
+    else if (tree.symbol == defn.DynamicTuple_dynamicToArray) transformTupleToArray(tree)
+    else super.transformApply(tree)
+  }
+
+  private def transformTupleCons(tree: tpd.Apply)(implicit ctx: Context): Tree = {
+    val TypeApply(_, headType :: tailType :: Nil) = tree.fun
+    val tail :: head :: Nil = tree.args
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/transform/TupleOptimizations.scala]: I find it very confusing that the arguments of a dynamicCons are reversed. Can we not leave them in the order head/tail?

File: compiler/src/dotty/tools/dotc/transform/TupleOptimizations.scala [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+    tupleTypes(tree.tpe) match {
+      case Some(tpes) =>
+        // Generate a the tuple directly with TupleN+1.apply
+        val size = tpes.size
+        if (size <= 5) {
+          // val t = tail
+          // TupleN+1(head, t._1, ..., t._n)
+          evalOnce(Typed(tail, TypeTree(defn.tupleType(tpes.tail)))) { tup =>
+            val elements = head :: (0 until size - 1).map(i => tup.select(nme.selectorName(i))).toList
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/transform/TupleOptimizations.scala]: The idiom
```
(0 until size - 1).map(i => tup.select(nme.selectorName(i))).toList
```
is used 4 times. Better to factor it out.

File: compiler/src/dotty/tools/dotc/transform/TupleOptimizations.scala [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+      val elements = (0 until size).map(_ => it.select(nme.next)).toList
+      knownTupleFromElements(tpes, elements)
+    } else {
+      // No optimization, keep:
+      // TupleXXL.fromIterator(it)
+      ref(defn.TupleXXL_fromIterator).appliedTo(it)
+    }
+  }
+
+  private def tupleTypes(tp: Type, bound: Int = Int.MaxValue)(implicit ctx: Context): Option[List[Type]] = {
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/transform/TupleOptimizations.scala]: There is already a `tupleTypes` in Definitions. Should that be deleted?

File: compiler/src/dotty/tools/dotc/transform/TupleOptimizations.scala (hunk 2 above)
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/transform/TupleOptimizations.scala]: The order is purely historical. I will reverse them.

**Comment by nicolasstucki** [comment]
Rebased

File: compiler/src/dotty/tools/dotc/transform/TupleOptimizations.scala [hunk 5]
```diff
@@ -24,7 +24,8 @@ class TupleOptimizations extends MiniPhase with IdentityDenotTransformer {
   def phaseName: String = "genericTuples"
 
   override def transformApply(tree: tpd.Apply)(implicit ctx: Context): tpd.Tree = {
-    if (tree.symbol == defn.DynamicTuple_dynamicCons) transformTupleCons(tree)
+    if (!tree.symbol.exists || tree.symbol.owner != defn.DynamicTupleModuleClass) super.transformApply(tree)
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/transform/TupleOptimizations.scala]: You don't need a `super.transformApply` in a MiniPhase. Just return `tree` itself.

Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/transform/TupleOptimizations.scala]: modified

**Pull request merged**